rust/src/util/types.rs

```diff
@@ -1,25 +1,115 @@
-use std::ptr;
+use std::{ops::Deref, ptr};
 
 use drop_struct_macro_derive::DropStructMacro;
 // `CodeAndMessage` is the trait implemented by `code_and_message_impl
 use ffi_toolkit::{code_and_message_impl, free_c_str, CodeAndMessage, FCPResponseStatus};
 
 #[repr(C)]
-#[derive(DropStructMacro)]
+pub struct fil_Array<T: Sized> {
+    ptr: *mut T,
+    len: usize,
+}
+
+impl<T> Default for fil_Array<T> {
+    fn default() -> Self {
+        Self {
+            ptr: ptr::null_mut(),
+            len: 0,
+        }
+    }
+}
+
+pub type fil_Bytes = fil_Array<u8>;
+
+impl<T> From<Vec<T>> for fil_Array<T> {
+    fn from(buf: Vec<T>) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        buf.into_boxed_slice().into()
+    }
+}
+
+impl<T> From<Box<[T]>> for fil_Array<T> {
+    fn from(buf: Box<[T]>) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        let len = buf.len();
+        let ptr = buf.as_mut_ptr();
+        Box::leak(buf);
```

Comment:
Are those to lines the same as `let ptr = Box::into_raw(buf)`?

Reply:
Not quite. `Box::into_raw(buf)` will return `*const [T]` instead of `*const T`.

Reply:
Thanks, I knew I must have missed something.